pfio/v2/fs.py

```diff
@@ -0,0 +1,381 @@
+import abc
+import contextlib
+import copy
+import os
+import stat
+from abc import abstractmethod
+from io import IOBase
+from types import TracebackType
+from typing import Any, Callable, Iterator, Optional, Type
+from urllib.parse import urlparse
+
+
+class FileStat(abc.ABC):
+    """Detailed file or directory information abstraction
+
+    :meth:`pfio.IO.stat` of filesystem/container handlers return an object of
+    subclass of ``FileStat``.
+    In addition to the common attributes that the ``FileStat`` abstract
+    provides, each ``FileStat`` subclass implements some additional
+    attributes depending on what information the corresponding filesystem or
+    container can handle.
+    The common attributes have the same behavior despite filesystem or
+    container type difference.
+
+    Attributes:
+        filename (str):
+            Filename in the filesystem or container.
+        last_modifled (float):
+            UNIX timestamp of mtime. Note that some
+            filesystems or containers do not have sub-second precision.
+        mode (int):
+            Permission with file type flag (regular file or directory).
+            You can make a human-readable interpretation by
+            `stat.filemode <https://docs.python.org/3/library/stat.html#stat.filemode>`_.
+        size (int):
+            Size in bytes. Note that directories may have different
+            sizes depending on the filesystem or container type.
+    """     # NOQA
+    filename = None
+    last_modified = None
+    mode = None
+    size = None
+
+    def isdir(self):
+        """Returns whether the target is a directory, based on the permission flag
+
+        Returns:
+            `True` if directory, `False` otherwise.
+        """
+        return bool(self.mode & 0o40000)
+
+    def __str__(self):
+        return '<{} filename="{}" mode="{}">'.format(
+            type(self).__name__, self.filename, stat.filemode(self.mode))
+
+    def __repr__(self):
+        return str(self.__str__())
+
+
+class ForkedError(RuntimeError):
+    pass
+
+
+class FS(abc.ABC):
+    '''FS access abstraction
+
+    '''
+
+    cwd = None
+    pid = os.getpid()
+
+    def __init__(self):
+        self.pid = os.getpid()
+
+    @abstractmethod
+    def open(self, file_path: str, mode: str = 'rb',
+             buffering: int = -1, encoding: Optional[str] = None,
+             errors: Optional[str] = None,
+             newline: Optional[str] = None,
+             closefd: bool = True,
+             opener: Optional[Callable[
+                 [str, int], Any]] = None) -> Type["IOBase"]:
+        raise NotImplementedError()
+
+    def open_zip(self, file_path: str, mode='r') -> Type["Zip"]:  # NOQA
+        from .zip import Zip
+        return Zip(self, file_path, mode)
+
+    def subfs(self, rel_path: str) -> Type["FS"]:
```

Comment:
Is there any reason we do not use the `chdir` name? Does it support switching from zip to normal directory?

Reply:
It's rather like chroot than chdir. It doesn't support going up the directory tree or switching the type. See also: https://golang.org/pkg/io/fs/#SubFS